pkg/prometheus/promcfg.go

```diff
@@ -1120,7 +1120,11 @@ func (cg *ConfigGenerator) generateServiceMonitorConfig(
 		role = kubernetesSDRoleEndpointSlice
 	}
 
-	cfg = append(cfg, cg.generateK8SSDConfig(m.Spec.NamespaceSelector, m.Namespace, apiserverConfig, store, role, nil))
+	if m.Spec.AttachMetadata != nil {
+		cfg = cg.WithMinimumVersion("2.37.0").AppendMapItem(cfg, "", cg.generateK8SSDConfig(m.Spec.NamespaceSelector, m.Namespace, apiserverConfig, store, role, m.Spec.AttachMetadata))
```

Comment:
hmm it would mean that the service monitor would generate no config at all if Prometheus < v2.37.0. We could rather extend `generateK8SSDConfig()` to pass the minimum version supporting node metadata (e.g. 2.35.0 for pod monitor and 2.37.0 for service monitor).

Reply:
Indeed, when I did this I did not realize it would only generate config if Prometheus > v2.37.0. However instead of extending the `generateK8SSDConfig()` I tried encapsulating the details of attachMetadata into a new type that we then pass to `generateK8SSDConfig()`. This to me makes slightly more sense but I'm up to just passing the minimum version if you feel it's better :)

Reply:
sorry for the lag on re-reviewing this. My suggestion was to actually remove the version check from `generateK8SSDConfig` and have the calling function responsible for validation prior with the applicable version check.

Simons suggestion or your alternative works for me too though.

Reply:
Apologies at the time I miss understood it, I'm fine with implementing either :) tests should be fixed now